File: app/src/main/java/com/example/schedule/ui/LoginActivity.java
package com.example.schedule.ui;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.example.schedule.API.ApiHolder;
import com.example.schedule.API.NetworkService;
import com.example.schedule.POJO.OK_POJO.JsonResponse;
import com.example.schedule.R;
import com.example.schedule.SQL.SQLScheduleEditor;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginActivity extends AppCompatActivity {

    private EditText text;
    private Button button;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        text = findViewById(R.id.group);
        button = findViewById(R.id.getScheduleButton);
        button.setEnabled(false);
        text.addTextChangedListener(textChangedListener);
        text.setOnEditorActionListener(inputGroup);
    }

    private static final int VERSION = 1;
    private static final String ACCESS_TOKEN = "23a07a867ead26ba489f0b8dc7ab1c330f44a93dc72255e6bed322d5c5577fc2f0517665b03fcbfa794f795833";
    private Call<com.example.schedule.POJO.OK_POJO.Response> responseCall;

    public void onClick(View v) {

        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(text.getWindowToken(), 0);
        LoginDialogFragment fragment = new LoginDialogFragment(() -> responseCall.cancel());
        fragment.show(getSupportFragmentManager(), null);
        String editText = text.getText().toString();

        responseCall = NetworkService
                .getInstance()
                .getJSONApi()
                .getSchedule(VERSION, ACCESS_TOKEN, editText);

        new Thread(() -> {
            responseCall
                    .enqueue(new Callback<com.example.schedule.POJO.OK_POJO.Response>() {
                        @Override
                        public void onResponse(@NonNull Call<com.example.schedule.POJO.OK_POJO.Response> call,
                                               @NonNull Response<com.example.schedule.POJO.OK_POJO.Response> response) {
                            com.example.schedule.POJO.OK_POJO.Response r = response.body();
                            JsonResponse jr = r.getJsonResponse();
                            if (r.getError() != null) {
                                switch (r.getError().getErrorCode()) {

                                    case ApiHolder.UNKNOWN_GROUP:
                                        Toast.makeText(getApplicationContext(), R.string.unknownGroup, Toast.LENGTH_LONG).show();
                                        text.setText("");

                                        break;
                                    default:
                                        Toast.makeText(getApplicationContext(), R.string.internalErrorOccurred, Toast.LENGTH_LONG).show();
                                }
                            } else if (jr != null) {
                                String tableName = editText.replace("-", "");
                                SQLScheduleEditor editor = new SQLScheduleEditor(getApplicationContext(),tableName,1);
                                editor.fillDataBase(jr);
                                editor.close();
                                Intent intent = new Intent(getApplicationContext(), MainActivity.class);
                                startActivity(intent);
                                fragment.dismiss();
                                finish();
                            }
                            fragment.dismiss();
                        }

                        @Override
                        public void onFailure(@NonNull Call<com.example.schedule.POJO.OK_POJO.Response> call,
                                              @NonNull Throwable t) {

                            if (!call.isCanceled()) {
                                Toast.makeText(getApplicationContext(), R.string.connectionError, Toast.LENGTH_LONG).show();
                            }
                            fragment.dismiss();
                        }
                    });
        }).start();
    }

    interface RequestOperation{
        void cancelRequest();
    }

    private TextWatcher textChangedListener = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable s) {
            if (s.toString().equals("")) {
                button.setEnabled(false);
            } else {

                button.setEnabled(true);
            }
        }
    };

    private TextView.OnEditorActionListener inputGroup = (v, actionId, event) -> {
        if (actionId == EditorInfo.IME_ACTION_DONE && button.isEnabled()) {

            onClick(null);
            button.setEnabled(false);

            return true;
        }
        return false;
    };
}

File: app/src/main/java/com/example/schedule/ui/LoginActivity.kt
package com.example.schedule.ui

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.schedule.R

class LoginActivity : AppCompatActivity(), LoginLoadingView {

    private var text: EditText? = null
    private var button: Button? = null
    private lateinit var presenter: LoginActivityPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        text = findViewById(R.id.group)
        button = findViewById(R.id.getScheduleButton)
        button!!.isEnabled = false
        text!!.addTextChangedListener(textChangedListener)
        text!!.setOnEditorActionListener(inputGroup)
        presenter = LoginActivityPresenter(this,applicationContext)
    }

    fun onClick(v: View?) {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(text!!.windowToken, 0)
        showLoading()
        val editText = text!!.text.toString()
        Thread(Runnable{presenter.makeRequest(editText)}).start()
    }

    override fun showGroupIsNotExists() {
        Toast.makeText(applicationContext, R.string.unknownGroup, Toast.LENGTH_LONG).show()
        text!!.setText("")
    }

    override fun showConnectionError() {
        Toast.makeText(applicationContext, R.string.connectionError, Toast.LENGTH_LONG).show()
    }

    override fun showInternalError() {
        Toast.makeText(applicationContext, R.string.internalErrorOccurred, Toast.LENGTH_LONG).show()
    }

    private val fragment = LoginDialogFragment(DialogInterface.OnCancelListener{
        presenter.cancelRequest()
    })

    override fun showLoading() {
        fragment.show(supportFragmentManager, null)
    }

    override fun hideLoading(toFinish : Boolean) {
        fragment.dismiss()
        if (toFinish) finish()

    }

    override fun startMainActivity() {
        startActivity(Intent(applicationContext,MainActivity::class.java))
    }

    private val textChangedListener: TextWatcher = object : TextWatcher {

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}

        override fun afterTextChanged(s: Editable) {
            button!!.isEnabled = s.toString() != ""
        }
    }

    private val inputGroup = OnEditorActionListener { _: TextView?, actionId: Int, _: KeyEvent? ->
        if (actionId == EditorInfo.IME_ACTION_DONE && button!!.isEnabled) {
            onClick(null)
            button!!.isEnabled = false
            true
        }else false
    }

    companion object {
        private const val VERSION = 1
        private const val ACCESS_TOKEN = "23a07a867ead26ba489f0b8dc7ab1c330f44a93dc72255e6bed322d5c5577fc2f0517665b03fcbfa794f795833"
    }
}
